### PR TITLE
ci: Generalize the GitLab CI pipeline yaml

### PR DESCRIPTION
**Pull request by chuckatkins**
This extracts the yaml refactoring from #22179 as a stand-alone change

### CODE REVIEW

File: share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml]: What do you think of changing this path component to `gitlab_cloud_pipelines` (or something) and renaming the directory these files live in?

File: share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml:75]: This is so nice!  Does it make sense to add an extra doc line here reminding developers to put their stack into a directory of the same name?  e.g.:

```
# Add super cool stack to spack repo:
# 
#     share/spack/gitlab/gitlab_cloud_pipelines/stacks/${SPACK_CI_STACK_NAME}/spack.yaml
```

**Review by scottwittenburg — DISMISSED** [review]
This is such a huge improvement, thank you @chuckatkins! :clap: :tada: 

I have a couple suggestions, but they're very minor, and I'm fine with this as-is.

File: share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml
Comment (chuckatkins) [reply share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml]: I like it.  I didn't want to change any file locations though in case it broke something e4s was doing.  It definitely makes sense to generalize the directory name too though.  I'll update accordingly.

File: share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml
Comment (chuckatkins) [reply share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml:75]: Certainly

**Comment by chuckatkins** [comment]
This will have a follow-up PR to remove `share/spack/gitlab/cloud_e4s_pipelines` once this is merged and the gitlab project setting are changed.

**Comment by chuckatkins** [comment]
Holding off merging until after the weekend just in case we break pipelines.

**Review by scottwittenburg — APPROVED** [review]

**Pull request merged**
